src/realm/object-store/impl/list_notifier.cpp

```diff
@@ -64,6 +64,20 @@ bool ListNotifier::do_add_required_change_info(TransactionChangeInfo& info)
     info.lists.push_back({m_table, m_obj.value, m_col.value, &m_change});
 
     m_info = &info;
+
+    // When adding or removing a callback, the related tables can change due to the way we calculate related tables
+    // when key path filters are set, hence we need to recalculate every time the callbacks are changed.
+    // We only need to do this for lists that link to other lists. Lists of primitives cannot have related tables.
+    if (m_did_modify_callbacks && m_type == PropertyType::Object) {
```

Comment:
Accessing this (and the callbacks to build the filtered set) requires locking the callback mutex.

Reply:
Fixed with 0775bb0a3.
Also added to all other `do_add_required_change_info`. For this change I've also extracted the calculation of `m_key_path_arrays` to whenever `m_callback` is updated. It's then only accessed instead of recalculated in `get_modification_checker`.